### ISSUE TITLE
incorrect TypeScript definitions for TelephoneNumber, Extension, PhoneCode

### ISSUE THREAD

**Issue by atsu85** [catamphetamine/libphonenumber-js#170]
_I'll send a PR shortly, just created this issue to document the problem_

## Problem
Currently PhoneCode (and TelephoneNumber and Extension) [are defined using interface](https://github.com/catamphetamine/libphonenumber-js/blob/master/index.d.ts#L5-L7), like:
`export interface PhoneCode extends String { }`
meaning, that variables/fields of these types are treated as objects, not string primitives
(even tough `typeof phoneCode === "string"`).

As a result variable of type PhoneCode can't be assigned to variables/fields with string type:
```TypeScript
import { getPhoneCode } from 'libphonenumber-js';

const phoneCode = getPhoneCode('EE');
const phoneCodeAsString: string = phoneCode;
// ERROR:^^^^^^^^^^^ TS2322: Type 'PhoneCode' is not assignable to type 'string'.
```

## Expected behaviour
Should be able to asign PhoneCode (and TelephoneNumber and Extension) to ´string´.

## Solution
Declare PhoneCode (and TelephoneNumber and Extension) using 
`export type XXX = string`
instead of
`export interface PhoneCode extends String { }`

--------------
Just in case, I'm mentioning @RonNewcomb, because he created the initial TypeScript definitions file, and also @catamphetamine who has created `Extension` type by following example of @RonNewcomb

**Comment by catamphetamine** [comment]
good job

**Comment by atsu85** [comment]
thanks for merged PR and maintaining this library :)

**Comment by RonNewcomb** [comment]
The assignment needs a cast:
`const phoneCodeAsString: string = <string>phoneCode;`

The problem with Typescript's `export type XXX = string` syntax is that the Intellisense of every IDE I've ever seen will show a var/property of type XXX as being of type string, not XXX.  That largely defeats the purpose of typing it at all. 

The reason that requiring a cast is (IMHO) a good thing is such:   If you have an Address1 type descended from string and a CustomerName type descended from string, and you assign a value of one to a property/var of the other, you *want* to get the compiler error that requires a cast, because they're different types despite both being represented as string.

This problem is particularly acute with string manipulation libraries like this one.  Strings in, strings out, did this string already go through the library or not, does the phone# format that that string is in match the format that this string function accepts... everything's a damn string. You're effectively working with untyped data since `string` is about as helpful in this context as `any`.  

Making subtypes that appear in intellisense helps a lot with simple usability, and the occasional required cast, albeit annoying, can save a you a bug hunt or two next week.  

But yeah I really wish Typescript / Intellisense was a bit better here so it would be a non-issue.

**Comment by atsu85** [comment]
@RonNewcomb, good point - it is a matter of trade-offs. According to my change it was possible to just document the type, allowing to assign it to string without casting, but the downside is that it won't prevent me from assigning one kind of string to different kind of string either.

I prepared a PR that reverts my previous change.

**Comment by atsu85** [comment]
@catamphetamine, sorry for the inconvenience - please merge my PR that reverts my previous change related to this issue

**Comment by catamphetamine** [comment]
Released `libphonenumber-js@1.0.12`

**Comment by catamphetamine** [comment]
@atsu85 Thx for trying
@RonNewcomb Thanks for clearing things out

**Comment by jineshshah36** [comment]
I would like to suggest potentially reopening this issue. Idiomatic Typescript favors just using the "string" type instead of extending "String". Most people using Typescript will expect the simple string type. Extending String should be left to the user and not required for everyone having to use the library. In most cases, the rest of the program will expect to consume the value as a regular string type. You are making a big assumption that people need to know the difference between CountryCode and TelephoneNumber at the type level. If they do, it should be on them to enforce that.

**Comment by jineshshah36** [comment]
I'm happy to make the changes if we can agree

**Comment by tduyduc** [comment]
Is it possible to use `type E164Number = string & {};` instead of `interface E164Number extends String { }`?

![image](https://user-images.githubusercontent.com/15998902/147766078-b8830139-4334-47eb-b742-5559ae2f5fe3.png)

The variables `s1` and `s2` will be named nominally (`String1` and `String2`), but there is a caveat: `(string & {}) extends object`, so a type assertion (or a variable with `string` type) might still be necessary.

**Comment by catamphetamine** [comment]
FYI: I'm about to merge [this new pull request](https://gitlab.com/catamphetamine/libphonenumber-js/-/merge_requests/8/diffs) that replaces the declaration for all "basic string" types like `PhoneCode`, `Extension`, etc.

Was:

```ts
export interface E164Number extends String { }
export interface NationalNumber extends String { }
export interface Extension extends String { }
export interface CarrierCode extends String { }
export interface CountryCallingCode extends String { }
```

Will become:

```ts
type Tagged<A, T> = A & { __tag: T };

type E164Number = Tagged<string, "E164Number">;
type NationalNumber = Tagged<string, "NationalNumber">;
type Extension = Tagged<string, "Extension">;
type CarrierCode = Tagged<string, "CarrierCode">;
type CountryCallingCode = Tagged<string, "CountryCallingCode">;
```

If there're any objections then post them here.
I'll wait for a couple of days before merging that pull request.

**Comment by tduyduc** [comment]
If we don't ever assign a `NationalNumber` variable to an `E164Number` one, then these opaque types are actually the better way, so no objections from me.

**Comment by catamphetamine** [comment]
> If we don't ever assign a NationalNumber variable to an E164Number one

No one should be assigning a "national (significant) number" to an E.164 because those're two different concepts.
Ok, thx.

**Comment by catamphetamine** [comment]
Published `libphonenumber-js@1.9.48`.

**Comment by atsu85** [comment]
@catamphetamine, cool workaround!
However i have one proposal for change, since as i understand [the change](https://gitlab.com/catamphetamine/libphonenumber-js/-/merge_requests/8) only involves type definitions, not values, then instead of this:
```TypeScript
type Tagged<A, T> = A & { __tag: T };
```
better option would be this:
```TypeScript
type Tagged<A, T> = A & { __tag?: T };
```
as former doesn't give TS compiler error when trying to assign `__tag` to `string`
```TypeScript
var e164: E164Number = "e164" as any
var tag: string = nationalNumber.__tag
```
but latter would, as expeted, because `__tag` will always be `undefined` instead of non-null/undefined `string`
